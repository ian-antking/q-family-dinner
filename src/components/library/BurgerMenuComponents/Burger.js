import React from 'react';
import { bool, func } from 'prop-types';
import styled from 'styled-components';

const StyledBurger = styled.button`
  padding: 0;
  width: 55px;
  height: 50px;
  background: none;
  border: none;
  cursor: pointer;
  display: flex;
  flex-flow: column;
  justify-content: space-evenly;
  align-items: center;

  &:focus {
    outline: none;
  }

  div {
    margin: 0;
    padding: 0;
    width: 45px;
    height: 8px;
    background: #BDBDBD;
    border-radius: 5px;
    transition: all 0.3s linear;
    transform-origin: 2.25px;

    :first-child {
      transform: ${({ open }) => (open ? 'rotate(45deg)' : 'rotate(0)')};
    }

    :nth-child(2) {
      opacity: ${({ open }) => (open ? '0' : '1')};
      transform: ${({ open }) => (open ? 'translateX(20px)' : 'translateX(0)')};
    }

    :nth-child(3) {
      transform: ${({ open }) => (open ? 'rotate(-45deg)' : 'rotate(0)')};
    }
  }
`;

const Burger = ({ open, setOpen }) => {
  return (
    <StyledBurger data-testid="burger" open={open} onClick={() => setOpen(!open)}>
      <div data-testid="slice-one"/> 
      <div data-testid="slice-two"/>
      <div data-testid="slice-three"/>
    </StyledBurger>
  );
};

Burger.propTypes = {
  open: bool.isRequired,
  setOpen: func.isRequired,
};

export default Burger;
