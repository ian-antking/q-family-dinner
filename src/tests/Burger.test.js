import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Burger from '../components/library/BurgerMenuComponents/Burger';

const setOpen = jest.fn();

test('snapshot: Burger component', () => {
  const { asFragment } = render(<Burger setOpen={setOpen} />);
  const component = asFragment();
  expect(component).toMatchSnapshot();
});

test('burger component renders', () => {
  const { getByTestId } = render(<Burger setOpen={setOpen} />);
  const component = getByTestId('burger');
  expect(component).toBeInTheDocument();
});

test('the function passed to onClick is called on click', () => {
  const { getByTestId } = render(<Burger setOpen={setOpen} />);
  const component = getByTestId('burger');
  fireEvent.click(component);
  expect(setOpen).toHaveBeenCalled();
});

xtest('when open = false the styles of child divs are such that they form a burger', () => {
  const { getByTestId } = render(<Burger setOpen={setOpen} />);
  const slice1 = getByTestId('slice-one');
  const slice2 = getByTestId('slice-two');
  const slice3 = getByTestId('slice-three');
  expect(slice1).toHaveStyle('transform: rotate(0)');
  expect(slice2).toHaveStyle(`
    opacity: 1;
    transform: translateX(0);
  `);
  expect(slice3).toHaveStyle('transform: rotate(0)');
});

xtest('when open = true the styles of child divs are such that they form a cross', () => {
  const { getByTestId } = render(<Burger open setOpen={setOpen} />);
  const slice1 = getByTestId('slice-one');
  const slice2 = getByTestId('slice-two');
  const slice3 = getByTestId('slice-three');
  expect(slice1).toHaveStyle('transform: rotate(45deg)');
  expect(slice2).toHaveStyle(`
  opacity: 0;
  transform: translateX(20px);
  `);
  expect(slice3).toHaveStyle('transform: rotate(-45deg)');
});
