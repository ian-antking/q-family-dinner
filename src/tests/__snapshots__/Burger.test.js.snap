// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: Burger component 1`] = `
<DocumentFragment>
  <button
    class="sc-AxjAm dOvMlp"
    data-testid="burger"
  >
    <div
      data-testid="slice-one"
    />
    <div
      data-testid="slice-two"
    />
    <div
      data-testid="slice-three"
    />
  </button>
</DocumentFragment>
`;
